src/blockchain/alt_block_tree.cpp

```diff
@@ -120,18 +120,51 @@ void AltTree::removePayloads(const AltBlock& containingBlock,
   assert(index);
 
   cmp_.removePayloads(*index, payloads);
+
+  if (index->containingContext.back().vbk.empty() &&
+      index->containingContext.back().vtbs.empty()) {
+    index->containingContext.pop_back();
+  }
 }
 
 template <>
 bool AltTree::PopForkComparator::sm_t::applyContext(
     const BlockIndex<AltBlock>& index, ValidationState& state) {
   return tryValidateWithResources(
       [&]() -> bool {
-        // apply vbk blocks after that
-        auto& vbk = index.containingContext.vbk;
-        for (size_t i = 0, size = vbk.size(); i < size; i++) {
-          if (!tree().acceptBlock(vbk[i], state)) {
-            return state.addIndex(i).Invalid("bad-vbk-block");
+        if (index.containingContext.empty()) {
+          return true;
+        }
+
+        // step 1
+        for (const auto& b : index.containingContext.back().vbk) {
```

Comment:
instead of using `index.containingContext.back()` every time, do a rename instead
```
auto& ctx = index.containingContext.back();
for(... ctx) {
```

Reply:
fixed